deltacat/io/redis_object_store.py

```diff
@@ -0,0 +1,114 @@
+import logging
+from ray import cloudpickle
+import time
+from deltacat.io.object_store import IObjectStore
+from typing import Any, List
+from deltacat import logs
+import uuid
+import socket
+import redis
+from collections import defaultdict
+
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+class RedisObjectStore(IObjectStore):
+    """
+    An implementation of object store that uses Redis in memory DB.
+    """
+
+    def __init__(self) -> None:
+        self.client_cache = {}
+        self.current_ip = None
+        self.SEPARATOR = "_"
+        super().__init__()
+
+    def put(self, obj: object, *args, **kwargs) -> Any:
+        serialized = cloudpickle.dumps(obj)
+        uid = uuid.uuid4()
+        current_ip = self._get_current_ip()
+        ref = self._create_ref(uid, current_ip)
+
+        client = self._get_client_by_ip(current_ip)
+        if client.set(uid.__str__(), serialized):
+            return ref
+        else:
+            raise RuntimeError(f"Unable to write {ref} to cache")
+
+    def put_many(self, objects: List[object], *args, **kwargs) -> List[Any]:
+        input = {}
+        result = []
+        current_ip = self._get_current_ip()
+        for obj in objects:
+            serialized = cloudpickle.dumps(obj)
+            uid = uuid.uuid4()
+            ref = self._create_ref(uid, current_ip)
+            input[ref] = serialized
+            result.append(ref)
+
+        client = self._get_client_by_ip(current_ip)
+
+        if client.mset(input):
+            return result
+        else:
+            raise RuntimeError("Unable to update cache")
+
+    def get_many(self, refs: List[Any], *args, **kwargs) -> List[object]:
+        result = []
+        uid_per_ip = defaultdict(lambda: [])
+
+        start = time.monotonic()
+        for ref in refs:
+            uid, ip = ref.split(self.SEPARATOR)
+            uid_per_ip[ip].append(uid)
+
+        for (ip, uids) in uid_per_ip.items():
+            client = self._get_client_by_ip(ip)
+            cache_result = client.mget(uids)
+            assert len(cache_result) == len(
+                uids
+            ), "Not all values were returned from cache"
+
+            total_bytes = 0
+
+            deserialize_start = time.monotonic()
+            for serialized in cache_result:
+                deserialized = cloudpickle.loads(serialized)
```

Comment:
This assumes that the object is cloudpickled first, which may not be true

Reply:
you can only get what you put. So, from the `put` and `put_many` implementation of `RedisObjectStore`, it should be always true.

Reply:
what if later we use this object store interface to put un-pickled object?

Reply:
That will be part of the implementation whether to pickle or not. The `object_store.get` will return what was `object_store.put`